CI: add retry to Calypso E2E and Gutenberg E2E specs.

worldomonation opened the pull request:
Related to pciE2j-2cY-p2.

## Proposed Changes

This PR adds a single retry of only failed tests.

## Testing Instructions

Locally, I modified an existing test to fail intentionally on first pass by throwing an exception halfway through.
If the environment variable `RETRY` is defined, then it will not throw an exception, allowing the step to pass.

I ran the first pass and it failed as expected.
Then, on the second pass, I add the `--onlyFailures` flag and defined the env var `RETRY` to allow the step to pass.
The second pass:
- only ran the failing test suite, as expected
- passed all steps

<details>
    <summary>Logs</summary>

```
➜  ~/workspace/wp-calypso-second/test/e2e git:(ci/e2e-add-retry) ✗ yarn jest test/e2e/specs/infrastructure/infrastructure__ssr-enabled.ts test/e2e/specs/me/me__smoke.ts
  console.log
    Intentionally failing...

      at Object.<anonymous> (specs/infrastructure/infrastructure__ssr-enabled.ts:23:15)

Artifacts for infrastructure__ssr-enabled: /Users/edwintakahashi/workspace/wp-calypso-second/test/e2e/results/infrastructure__ssr-enabled__2023-06-22_09-08-44-745-GmNF2y
 FAIL  specs/infrastructure/infrastructure__ssr-enabled.ts
  Server-side Rendering
    Check SSR endoint: log-in
      ✓ Check SSR: log-in (2378 ms)
    Check SSR endoint: themes
      ✕ Check SSR: themes (143 ms)
    Check SSR endoint: theme/twentytwentythree
      ○ skipped Check SSR: theme/twentytwentythree

  ● Server-side Rendering › Check SSR endoint: themes › Check SSR: themes



      22 | 					if ( endpoint === 'themes' ) {
      23 | 						console.log( 'Intentionally failing...' );
    > 24 | 						throw new Error();
         | 						      ^
      25 | 					}
      26 | 				}
      27 | 				await page.goto( DataHelper.getCalypsoURL( endpoint ), { timeout: 20 * 1000 } );

      at Object.<anonymous> (specs/infrastructure/infrastructure__ssr-enabled.ts:24:13)

 PASS  specs/me/me__smoke.ts (7.452 s)
  Me: Smoke Test
    ✓ Navigate to /me (809 ms)
    ✓ Navigate to Me > Account Settings (79 ms)
    ✓ Navigate to Me > Purchases (164 ms)
    ✓ Navigate to Me > Security (248 ms)
    ✓ Navigate to Me > Privacy (85 ms)
    ✓ Navigate to Me > Notification Settings (114 ms)
    ✓ Navigate to Me > Blocked Sites (264 ms)
    ✓ Navigate to Me > Apps (129 ms)

Test Suites: 1 failed, 1 passed, 2 total
Tests:       1 failed, 1 skipped, 9 passed, 11 total
Snapshots:   0 total
Time:        8.215 s, estimated 11 s
Ran all test suites matching /test\/e2e\/specs\/infrastructure\/infrastructure__ssr-enabled.ts|test\/e2e\/specs\/me\/me__smoke.ts/i.
➜  ~/workspace/wp-calypso-second/test/e2e git:(ci/e2e-add-retry) ✗ RETRY=1 yarn jest test/e2e/specs/infrastructure/infrastructure__ssr-enabled.ts test/e2e/specs/me/me__smoke.ts --onlyFailures
 PASS  specs/infrastructure/infrastructure__ssr-enabled.ts (8.72 s)
  Server-side Rendering
    Check SSR endoint: log-in
      ✓ Check SSR: log-in (1971 ms)
    Check SSR endoint: themes
      ✓ Check SSR: themes (2326 ms)
    Check SSR endoint: theme/twentytwentythree
      ✓ Check SSR: theme/twentytwentythree (3635 ms)

Test Suites: 1 passed, 1 total
Tests:       3 passed, 3 total
Snapshots:   0 total
Time:        8.972 s
Ran all test suites matching /test\/e2e\/specs\/infrastructure\/infrastructure__ssr-enabled.ts|test\/e2e\/specs\/me\/me__smoke.ts/i.
```
</details>

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

worldomonation committed:
.teamcity/_self/lib/customBuildType/E2EBuildType.kt
- add a second command to retry failed tests.

dpasque commented:
@WunderBart 

> I'm assuming that if all tests passed in the first run, the second attempt would say nothing to test or something of that sort?

Yep! 🙂  It looks like this:
```
No failed test found.
Run Jest without `--onlyFailures` or with `--all` to run all tests.
```

dpasque commented:
@worldomonation Awesome!! Thank you SO much! 😄 This was exactly what I was thinking, and I really appreciate you knocking this out!

I think the one thing we might be missing is in the Build definition --> failureConditions --> We need to set `supportTestRetry` to `true` -- this helps the builds run green if multiple tests are run and have different outputs:
https://teamcity.jetbrains.com/app/dsl-documentation/root/failure-conditions/support-test-retry.html

worldomonation commented:
> @worldomonation Awesome!! Thank you SO much! 😄 This was exactly what I was thinking, and I really appreciate you knocking this out!
> 
> I think the one thing we might be missing is in the Build definition --> failureConditions --> We need to set `supportTestRetry` to `true` -- this helps the builds run green if multiple tests are run and have different outputs: https://teamcity.jetbrains.com/app/dsl-documentation/root/failure-conditions/support-test-retry.html

Thanks for the heads up, I will try adding that into the conditions. Once ready I'll ask for a re-review.

worldomonation committed:
_self/lib/customBuildType/E2EBuildType.kt
_self/projects/WebApp.kt
- add supportTestRetry flag to Kotlin DSL.

dpasque commented:
@worldomonation -- I took a look at the E2E runs in CI, and I noticed that it _should_ have done a retry, but it didn't!

I didn't realize this before, but after poking around in the Admin view, it looks like all TeamCity bash scripts get `set -o errexit` set by default, so the script will quit execution anytime it hits a non-zero code.

So, I think we have to put the second rerun into its own bash script step, right? Otherwise it will never run! Because if the first run fails, it will return a 1, which will exit the script before the rerun happens.

worldomonation commented:
> @worldomonation -- I took a look at the E2E runs in CI, and I noticed that it _should_ have done a retry, but it didn't!
> 
> I didn't realize this before, but after poking around in the Admin view, it looks like all TeamCity bash scripts get `set -o errexit` set by default, so the script will quit execution anytime it hits a non-zero code.
> 
> So, I think we have to put the second rerun into its own bash script step, right? Otherwise it will never run! Because if the first run fails, it will return a 1, which will exit the script before the rerun happens.
> 
> Sorry I missed that initially! 🙇

Hmm, that makes sense. 

Either we'd need a new step, or to `set +e` before the `yarn jest ...` part. 

[Reference](https://intoli.com/blog/exit-on-errors-in-bash-scripts/)

worldomonation committed:
.teamcity
- disable errexit prior to running the suite, and re-enable before the retry.

worldomonation commented on .teamcity/_self/lib/customBuildType/E2EBuildType.kt:
```diff
@@ -145,8 +145,17 @@ open class E2EBuildType(
 					mkdir temp
 
+					# Disable exit on error to support retries.
+					set +o errexit
```
@noahtallen @WunderBart @dpasque **comment:blocking** - while I understand there's nothing more permanent than a temporary solution, I am not comfortable adding a build step (and the associated overhead, however minor) to all Pre-Release Tests and Calypso E2E builds. 

What do you think? Would you both prefer to have a separate step for clear separation?

noahtallen replied:
My main question is just whether there is all that much overhead in an extra step -- it doesn't seem to take that much time in my experience, unless there is a different issue here.

That said, I don't think it matters that much. It's only really useful to have an extra step for people who debug the test suites a lot (e.g. us), so if you don't think it makes a big difference, I don't think we need to care about it too much

dpasque replied:
I'm fine either way too!

> it doesn't seem to take that much time in my experience, unless there is a different issue here.

This was my experience too, but again, I don't think it really matters whether we make a separate build step or toggle the flags like you have it 😄

worldomonation commented:
Looks like the retry is working as of the most recent revision:

<img width="1000" alt="image" src="https://github.com/Automattic/wp-calypso/assets/6549265/e2afaacb-4a9a-47ab-9b00-dcc2574cadc7">

[Check out this build for an example](https://teamcity.a8c.com/buildConfiguration/calypso_calypso_WebApp_Calypso_E2E_Playwright_desktop/10270345?buildTab=log&focusLine=0&linesState=245.855.856.857.860.923.924.927.933.1289.1290.1291.1343.1344.1345.1348&logView=flowAware).

worldomonation commented:
However, there is a new issue: if the same spec fails twice in a row, the artifact name from Run 2 clashes with the one from Run 1, causing `xargs -r -0 mv -t screenshots` to fail. This in turn results in the trace and log files not being copied.

<img width="1034" alt="image" src="https://github.com/Automattic/wp-calypso/assets/6549265/49bff895-ccd7-4d0c-bf18-cf573308ef11">

dpasque commented:
> However, there is a new issue: if the same spec fails twice in a row, the artifact name from Run 2 clashes with the one from Run 1, causing xargs -r -0 mv -t screenshots to fail. This in turn results in the trace and log files not being copied.

Oh shoot, I hadn't thought of that!

I think of the options you presented, I like number 2.

Really, I think what we want is just a way to...
1. Record artifacts for all failures, even if they pass in subsequent reruns (this helps with flaky debugging).
2. Easily distinguish for failing tests the artifacts from the first or second run.

I think time string is fine, or we could honestly just do something like have an ENV var called "RETRY_NUMBER" and when we do the rerun command, set `RETRY_NUMBER=1`, and then we could use that value in the artifact name (maybe append 'retry-1' or something). I think that would be pretty simple too, and would maybe make it a bit clearer when looking at the artifacts which is which. But that said, I think the time string is fine too to just get this rolling! 😄

worldomonation committed:
packages/calypso-e2e/src/jest-playwright-config/environment.ts
- add datetime to the artifact filename to bypass issue with the retry clashing the filename

worldomonation commented:
@dpasque Implemented fix in https://github.com/Automattic/wp-calypso/pull/78528/files#diff-d33147995e808c2f248624314bf232597704a415a95e3df79f653ea40f626a41R172.

worldomonation committed:
packages/calypso-e2e/src/jest-playwright-config/environment.ts
- change variable names to better reflect its use.
- apply the date-time string to all artifacts.
- remove logging and the resulting .log file, as it is rarely if ever used.

worldomonation commented:
Yet another problem: I think it would be nice to denote that a build had failures, but passed on retry. How could I do that?

Solution: a build tag is useful because it's entirely contained within TeamCity, is flexible and can be applied to any build. 

However, TeamCity being TeamCity, there isn't a way to do this unless one hacks a REST API call from a build step. You can see that I was able to post a tag on a still-running build:
![image](https://github.com/Automattic/wp-calypso/assets/6549265/99524992-556d-4545-a0ce-bbfd3b03d09e)

Sandbox build: https://teamcity.a8c.com/admin/editBuildRunners.html?id=buildType%3ATeamcitySandbox_EdwinTest

**What's the issue then?**
For prototyping, I think we can use a CLI token tied to my TeamCity username that I will generate. We don't know if this is required once Playwright Test Runner is in the picture, because that test runner likely does its own tracking of retries.

But, should my account on TeamCity be terminated for any reason, then the step will stop working until the access token is replaced with another one. So if possible, it would be best to create an access token using a Systems-managed bot user.

worldomonation commented on packages/calypso-e2e/src/jest-playwright-config/environment.ts:
```diff
@@ -110,20 +110,10 @@ class JestEnvironmentPlaywright extends NodeEnvironment {
 		// Start the browser.
 		const browser = await browserType.launch( {
 			...config.launchOptions,
-			logger: {
-				log: async ( name: string, severity: string, message: string ) => {
-					await fs.appendFile(
-						logFilePath,
-						`${ new Date().toISOString() } ${ process.pid } ${ name } ${ severity }: ${ message }\n`
-					);
-				},
-				isEnabled: ( name ) => name === 'api',
-			},
```
I've chosen to remove the capturing of `.log` files due to a persisting clash in the log filenames _if there was a retry_.

1. because the file capture is started as part of the browser launch, it has no context on whether the test step will pass/fail. 
2. if the build required a retry, it will generate a new log file with the same name.
3. content captured in the log file isn't very useful due to https://github.com/microsoft/playwright/issues/5871. Besides, trace files have supplanted debugging needs.

dpasque reviewed (approved):
Btw - dropping an approval here so you can merge when you feel fit! 👍 I'm good with all the changes and approaches you've laid out, and don't think we need to worry about tagging builds for the moment. I think we can just get something simple going that starts turning more builds green! 😄
